### DISCUSSION TITLE
How best to generate OpenAPI schema with reusable enums?

### DISCUSSION THREAD

**Discussion by calvinwyoung** [fastapi/fastapi#7544] [Questions] [answered]
### First check

* [X] I used the GitHub search to find a similar issue and didn't find it.
* [X] I searched the FastAPI documentation, with the integrated search.
* [X] I already searched in Google "How to X in FastAPI" and didn't find any information.

### Description

How can I generate an OpenAPI schema with reusable enum classes?

If I have 2 pydantic models that reference the same enum class. By default, the generated OpenAPI schema will look something like this:

```json
{
  "components": {
    "schemas": {
      "Foo": {
        "title": "Foo",
        "type": "object",
        "properties": {
          "some_enum": {
            "title": "Some Enum",
            "enum": ["VALUE_ONE", "VALUE_TWO"],
            "type": "string"
          }
        }
      },
      "Bar": {
        "title": "Bar",
        "type": "object",
        "properties": {
          "some_enum": {
            "title": "Some Enum",
            "enum": ["VALUE_ONE", "VALUE_TWO"],
            "type": "string"
          }
        }
      }
    }
  }
}
```

I'm looking for a way to generate something like this instead:

```json
{
  "components": {
    "schemas": {
      "Foo": {
        "title": "Foo",
        "type": "object",
        "properties": {
          "some_enum": {
            "$ref": "#/components/schemas/SomeEnum"
          }
        }
      },
      "Bar": {
        "title": "Bar",
        "type": "object",
        "properties": {
          "some_enum": {
            "$ref": "#/components/schemas/SomeEnum"
          }
        }
      },
      "SomeEnum": {
        "type": "string",
        "enum": ["VALUE_ONE", "VALUE_TWO"]
      }
    }
  }
}
```

The main use case for this feature is to facilitate client code generation — the first schema will produce 2 separate enum classes in the client code, whereas the second schema will generate a single enum class that's referenced in both the `Foo` and `Bar` models. This follows the "Reusable enums" pattern here:
https://swagger.io/docs/specification/data-models/enums/

I recognize that this is related to pydantic as well as fastapi, but given that fastapi has its own OpenAPI layer on top of pydantic, I was hoping to find a solution here.

Is there a recommended way to achieve this behavior?

**Comment by vadim-su** [comment]
I've the same problem. Did you find any solutions?

**Comment by calvinwyoung** [comment, marked as answer]
@deterok I opened this PR in the pydantic repo:
https://github.com/samuelcolvin/pydantic/pull/1432

This issue will be fixed once that PR gets merged in. Feel free to chime in on that PR if you need it sooner.

For now, I'll go ahead and close this issue since it's really more of a pydantic thing.

**Comment by tiangolo** [comment]
Thanks for reporting back and closing the issue @calvinwyoung! :bow: 

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
